### ISSUE TITLE
Upgrading to a jspm version that uses SystemJS 0.20.x ( jspm version >= 0.17.0-beta.33) 

### ISSUE THREAD

**Issue by HeinrichFilter** [jspm/jspm-cli#2368]
**JSPM Version:** 0.17.0-beta.45

**Transpiler Plugin(s):** babel

**Details:**
We've been stuck on  jspm 0.17.0-beta.32 because I couldn't get our code to work with newer jspm versions that use SystemJS 0.20.x

I'm trying again and would appreciate some help.

My main problem is the following breaking change mentioned in the [SystemJS 0.20.0 release notes](https://github.com/systemjs/systemjs/releases/tag/0.20.0)
> Removes the ability to use named imports from non-ES modules (eg import {readFile} from 'fs' should be import fs from 'fs'; fs.readFile). This is in order to align with the NodeJS ES module directions

I've got lots of code that looks like the this:
```javascript
import React, { Component } from "react";
import { List, Map, Set } from "immutable";
```

This code doesn't work with SystemJs 0.20.0. `Component`, `List`, `Map`, `Set` are all undefined. 

What is the recommended way to migrate this type of code?

Should we change all our files to look like this:
```javascript
import React from "react";
import Immutable from "immutable";

const { Component } = React 
const { List, Map, Set } = Immutable
```

Or is there an easier way?

**Comment by guybedford** [comment]
You can try adding:

```
System.config({
  meta: {
    'react': { esModule: true },
    'immmutable': { esModule: true }
  }
})
```

(or through package config)

So that named exports get supported for CommonJS packages. `{ meta: { '*': { esModule: true } } }` may be a blanket hack, but not advised :)

**Comment by HeinrichFilter** [comment]
Thanks @guybedford for your help!

Unfortunately though, I can't get it to work.

Here is a simple repo showing the problem: https://github.com/HeinrichFilter/jspm-systemjs-v20

`Component` still ends up `undefined` even after adding the recommended meta setting:
![screen shot 2017-10-09 at 11 14 01 pm](https://user-images.githubusercontent.com/265415/31358880-bce65c2c-ad47-11e7-866b-e8b6fc776f47.png)

**Comment by HeinrichFilter** [comment]
When I move the `meta` config into `package.json` I get a different error (see https://github.com/HeinrichFilter/jspm-systemjs-v20/commit/6f0fb90cca2fe625ce194f19d8fdfdb0ce11c75c):
```
Uncaught (in promise) Error: process is not defined
  Evaluatingvhttp://localhost:8000/jspm_packages/npm/react@16.0.0/index.js
  Evaluating http://localhost:8000/index.js
```

`jspm_packages/npm/react@16.0.0/index.js` looks like this:
```javascript
(function(System, SystemJS) {(function (require, exports, module, __filename, __dirname, global, GLOBAL) {'use strict';

if (process.env.NODE_ENV === 'production') {
  module.exports = require('./cjs/react.production.min.js');
} else {
  module.exports = require('./cjs/react.development.js');
}

}).apply(__cjsWrapper.exports, __cjsWrapper.args);
})(System, System);
```

I updated the `package.json` by running:
```
jspm install npm:react -o '{ meta:{ esModule: true }}'
```

**Comment by HeinrichFilter** [comment]
@guybedford (or anyone else) any other ideas how to fix this, please? There are quite a few bug fixes that we need in newer versions of JSPM

**Comment by aluanhaddad** [comment]
@HeinrichFilter that looks like a separate issue. Very problematic nevertheless. The quick fix would be to override the config pointing at that react index file to instead point at the file it requires.

**Comment by aluanhaddad** [comment]
Oddly, I was experimenting with react using beta 47 not two days ago and everything seems to be working fine with 16.0.2

**Comment by aluanhaddad** [comment]
@HeinrichFilter I got your repo to work and sent you a PR. You had an invalid override.

I think we should do a better job of detecting erroneous configuration in these scenarios based on likely user intent.

As for importing React, well it is a mess. The actual implementation now just re-exports a single `const`.

See [here](https://github.com/facebook/react/blob/825682390d825bb51be2e06835cf7e3458352e83/packages/react/index.js#L16), [here](https://github.com/facebook/react/blob/master/packages/react/src/React.js#L28), and especially [this perplexing comment](https://github.com/facebook/react/blob/master/packages/react/index.js#L14).

I honestly don't know how approach tooling in light of that comment. Since that has been shipping.

I would advise you to qualify all reference to members of the conceptual namespace `React`. So `React.Component` and so on since they have gone from CommonJS to a _very_ oddly structured ESM distribution

**Comment by HeinrichFilter** [comment]
Thanks for the effort and PR @ahmedshuhel!

Just to clarify, my original question wasn't just about React or Immutable but all the CommonJS libs that we use from npm. We have 100s of imports that do named imports from non-ES modules.

Should they all be manually converted from this style `import { List, Map, Set } from "immutable";` to this style `import Immutable from "immutable"; const { List, Map, Set } = Immutable;`? Or is there a quicker way?

If it's just React that's currently problematic then I don't mind doing the manual work. But if it is **all** cjs imports then I want to make double sure before making the updates.

**Comment by aluanhaddad** [comment]
For a long while now, Babel and TypeScript have been automatically add the following when transforming ES Modules to CommonJS Modules or AMD Modules
```js
Object.defineProperty(exports, "__esModule", {
    value: true
});
```
Some hand authored packages explicitly set this as well, notably tslib.
This is a source level marker that indicates the very same as the SystemJS configuration property
```json
"esModule": true
```
SystemJS looks for the `__esModule` flag and treats the module accordingly.

If the flag present with a value of `true`, then named imports are allowed from the module.

If the flag is not present or specified has a value of `false`, then the entirety of `module.exports` is lifted to a synthesized `default` for consumption by the importing code.

This complies with the requirements of the ES Module specification and the requirements for interop with Node.js's implementation of small subset of the ES Module functionality.

It is what enables code like the following
```js
module.exports = $;
```
to be consumed as 
```js
import $ from 'jquery';
```
and is indeed the behavior exhibited by React.

Simplified, the logic looks roughly like
```js
function createInteropModuleRecord({exports}) {
  if(exports.__esmodule === true) {
    return exports;
  } else {
    return {default: exports};
  }
}
```

**Comment by HeinrichFilter** [comment]
Thanks for the in-depth explanation. This aligns with what I saw last night.

I manually updated my React and Immutable imports (to use `React.Component` and `Immutable.Map` approach respectively) and now everything is working. I didn't have to do anything for all the other named imports coming in from npm.

I saw the other npm libraries had `"format": "esm"`-type config in their `jspm_packages` config files. I assume that config is built using the `__esModule` flag that you explained? Eitherway, thanks for your help.
